examples/AcidRain.hs

```diff
@@ -31,39 +32,46 @@ userAction = Stream.repeatM $ liftIO askUser
             "quit"   -> return Quit
             _        -> putStrLn "Type potion or harm or quit" >> askUser
 
-acidRain :: MonadAsync m => SerialT m Event
+acidRain :: (MonadIO m, MonadAsync m) => Stream m Event
 acidRain =
-      Stream.repeatM (liftIO $ return $ Harm 1) -- AsyncT m Event
-    & Stream.constRate 1                        -- AsyncT m Event
-    & Stream.fromAsync                          -- SerialT m Event
+    Stream.repeat (liftIO $ return $ Harm 1)      -- Stream m Event
+        & Concur.parSequence (Concur.maxRate 1)
 
 data Result = Check | Done
 
-runEvents :: (MonadAsync m, MonadState Int m) => SerialT m Result
-runEvents = do
-    event <- userAction `Stream.parallel` acidRain -- SerialT m Event
-    case event of
-        Harm n -> modify (\h -> h - n) >> return Check
-        Heal n -> modify (\h -> h + n) >> return Check
-        Quit -> return Done
+runEvents :: (MonadIO m, MonadAsync m) => IORef Int -> Stream m Result
+runEvents ref =
+    let events = Concur.eval $ Concur.append2 userAction acidRain
+        e2 = Stream.takeEndBy (== Quit ) events     -- Stream m Event
+      in Stream.mapM modCount e2                    -- Stream m Result
+
+    where
+
+    modCount ev =
+        case  ev of
+            Harm n -> liftIO $ modifyIORef ref (\h -> h - n) >> return Check
+            Heal n -> liftIO $ modifyIORef ref (+ n) >> return Check
+            Quit -> return Done
 
 data Status = Alive | GameOver deriving Eq
 
-getStatus :: (MonadAsync m, MonadState Int m) => Result -> m Status
-getStatus result =
-    case result of
-        Done  -> liftIO $ putStrLn "You quit!" >> return GameOver
-        Check -> do
-            h <- get
-            liftIO $ if (h <= 0)
-                     then putStrLn "You die!" >> return GameOver
-                     else putStrLn ("Health = " <> show h) >> return Alive
+getStatus :: MonadAsync m => IORef Int -> Result -> m Status
+getStatus ref result =
+    liftIO $
+        case result of
+            Done  -> putStrLn "You quit!" >> return GameOver
+            Check -> do
+                h <- readIORef ref
+                if h <= 0
+                        then putStrLn "You die!" >> return GameOver
+                        else putStrLn ("Health = " <> show h) >> return Alive
 
 main :: IO ()
 main = do
     putStrLn "Your health is deteriorating due to acid rain,\\
-             \ type \"potion\" or \"quit\""
+        \ type \"potion\" or \"quit\""
+    ref <- newIORef 60
     let runGame =
-            Stream.mapM getStatus runEvents -- SerialT (StateT Int IO) Status
-          & Stream.drainWhile (== Alive)    -- StateT Int IO ()
-    void $ runStateT runGame 60
+          Stream.mapM (getStatus ref) (runEvents ref)
+            & Stream.fold (Fold.takeEndBy (/= Alive) Fold.drain)
```

Comment:
Why can't we use StateT as the inner monad? Just use Stream.runStateT to run it.